### ISSUE TITLE
Provide two annotations, NonnulllFields and NonnullApi

### ISSUE THREAD

**Issue by Silentdoer** [jakartaee/common-annotations-api#124]
Like in Spring, non null annotations can be added to all fields in the class and method parameter return values, And it should be able to support annotation of the entire project/module

**Comment by xenoterracide** [comment]
I'd like to see the converse also added NullField and NullApi, this way you could enforce being explicit.

Honestly this is low hanging fruit, I have time on my plate and would be willing to contribute these, as well as any underlying needed duplication of JSR-305 to support them (I notice spring's annotations compose JSR-305 code)

**Comment by sdeleuze** [comment]
Hey, Spring Framework committer here. I am not sure that would be as simple as adding those 2 additional annotations since:
 - Spring use case in leveraging meta annotations 
 - The scope where null-safety applies needs to be specified, in JSR 305, we use `@TypeQualifierDefault`
 - Annotation attributes like `When.MAYBE` need to be expressed in another way in order to avoid warnings from `javac`
 - Need to have specifications and tooling support

In a nutshell, going beyond simple `@Nonnull` and `@Nullable` annotations is a huge task that is much more involved that it seems, as proven in https://github.com/jspecify/jspecify related work.

**Comment by hantsy** [comment]
Moving JSpecify is great.

**Comment by gavinking** [comment]
As I've commented on several other issues, this is inappropriate. Jakarta Annotations is a place for common annotations reused across several other Jakarta EE specs. It is not a place to throw a half-baked API for static code analysis.

**Comment by hantsy** [comment]
> It is not a place to throw a half-baked API for static code analysis.

The existing Nonnull and Nullable is not for static analysis?

**Comment by gavinking** [comment]
> The existing Nonnull and Nullable is not for static analysis?

Indeed:

1.  I believe they are intended for that, and
2. I also believe that, strictly-speaking, they probably should **not** have been added to Jakarta annotations, since, to the best of my knowledge, at present **no Jakarta specification actually uses them** nor even talks about them, and, furthermore, **this is an incredibly inappropriate place to start defining a static analysis tool**.

On the other hand, since they were already added, we can use them in the Persistence spec, which goes some way to justifying that they should have been added after all.

**Comment by gavinking** [comment]
I mean, @hantsy, if you look at the original issue which proposed adding them:

https://github.com/jakartaee/common-annotations-api/issues/89

I don't see any discussion of why it was appropriate to add them to Jakarta Annotations, or of how they relate to the Jakarta EE platform, or to other Jakarta EE specifications.

Perhaps I'm missing something.

**Comment by xenoterracide** [comment]
The entirety of spring boot 4 stack is moving to Jspecify. There is a problem with the Jakarta annotations in that they are not runtime available (at least in the version I've had. It causes problems with tools like spot bugs). I think now I would even go so far as to say that Jakarta should consider removing them in favor of Jspecify and start specifying jspecfies usage on certain standardized contracts.

**Comment by hantsy** [comment]
Currently, only see Persistence contains a PR to apply these annotations, not see other spec follows.

There are no build tools (Maven, Gradle APT extensions) to handle this. 

1. Static analysis of the code risk by NPE.
2. Build time improvement, such as adding NULL check to the classes files that equivalent to `Objects.requiresNonNull()` code in the class instantiation or first place in methods.

**Comment by gavinking** [comment]
> The entirety of spring boot 4 stack is moving to Jspecify.

I have quite literally no idea why it's at all relevant what Spring is doing.

Jakarta Annotations exists to service the needs of the Jakarta EE platform.

Spring have chosen, for decades, to _not_ implement Jakarta EE.

**Comment by gavinking** [comment]
> I think now I would even go so far as to say that Jakarta should consider removing them in favor of Jspecify

I've already answered this one: JSpecify is not part of Jakarta EE, nor of Java SE, and so no Jakarta specification is permitted to have a dependency on JSPecify.

> There is a problem with the Jakarta annotations in that they are not runtime available

Not sure what you mean. These annotations are defined as follows:

```java
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface Nonnull {
}
```
```java
@Documented
@Retention(RUNTIME)
public @interface Nullable {
}
```

Both are declared with `RUNTIME` retention.

**Comment by xenoterracide** [comment]
They didn't used to be declared as runtime. Pretty confident I have another issue open somewhere complaining about that or at least there was one. I know that at one point I encounter a problem with spotless.

 As far as spring I was just pointing out major platform (multiple of them) . As well as the fact that Jspecify actually provides a specification for how these work and are meant to be used. Tooling checks like null away and checker framework are working to implement said specification.

Of course you're welcome to do as you will but if we're there's an argument to remove the annotations at some point then there's a reasonable argument to still have annotations but use the JSpecify specification and annotations instead.

All in all fair enough. I've said my piece.
